Add AnaplanService wrapper for the plugins

niceyang opened the pull request:
Add AnaplanService wrapper for the plugins making connections with Anaplan server.

niceyang committed:
Add AnaplanService wrapper for the plugins making connections with Anaplan server.

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+package com.anaplan.client;
```
is this our code? If so it shouldn't use the com.anaplan prefix, and should have a copyright header.

albertshau replied:
or does it need to be here because required classes like Program are not public?

niceyang replied:
Yeah it's our code. And yes it has to be in this package because the required methods in Program are package-private.

The copyright header is added.

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+      Program.setServiceLocation(new URI(serviceLocation));
+      Program.setAuthServiceLocation(new URI(authServiceLocation));
+    } catch (URISyntaxException e) {
+      throw new RuntimeException(e.getMessage());
```
catching a checked exception and rethrowing as a RuntimeException is usually a sign that errors are not being handled in the right place. 

In this case, it's better if the AnaplanConfig's getter methods return URI instead of String, so that anything using AnaplanConfig doesn't have to handle invalid values. Parsing strings into URI should happen as early as possible, where a better error message can be thrown. For example, if these are user provided values, they should be validated at a higher level, before it reaches this level.

niceyang replied:
Thanks for the instruction! Updated!

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+   * @throws InterruptedException when errors happens during the function run
+   */
+  public static void runAnaplanFunction(
+      String workspaceId, String modelId, String functionId, FunctionType type)
```
our plugin projects use 2 space indentation (similar comment in other places)

niceyang replied:
Got, updated!

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+ * Wrapper class of the Anaplan service providing the interfaces for a client to connect with
+ * Anaplan server.
+ */
+public class AnaplanService implements Serializable {
```
does this need to be Serializable? It seems more like a utility class with only static methods

niceyang replied:
No, it's cleaned up.

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+   * @param modelId the ID of the target model
+   * @param functionId the ID of the function to be run
+   * @param type the type {@link FunctionType} of the function to be run
+   * @throws InterruptedException when errors happens during the function run
```
javadoc should include that it throws ProcessesNotFoundException and ActionsNotFoundException

niceyang replied:
Thanks! Updated!

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+   * @param modelId the ID of the target model
+   * @param fileId the ID of the placeholder in Anaplan model for the file to be loaded
+   * @param chunkSize the size of a chunk for uploading
+   * @return the {@link OutputStream} for uploading a data file
```
javadoc should mention that it can throw ServerFilesNotFoundException

niceyang replied:
Thanks, updated!

niceyang committed:
Address comments from Albert:
0. Update the getServiceLocation() and getAuthServiceLocation()
1. Adding the copy right header
2. Fix the format
3. Add Javacode with exceptions

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -26,9 +40,9 @@
   /** Interface for the CDF plugin that use Anaplan service. */
   public interface AnaplanConfig {
-    String getServiceLocation();
+    URI getServiceLocation() throws URISyntaxException;
```
this shouldn't throw URISyntaxException.  Parsing the String into a URI should happen before the AnaplanConfig is created, as we don't want callers to have to handle the exception.

niceyang replied:
Thanks for this point! Here comes the limitation ->

The AnaplanConfig will be implemented by PluginConfig. And the serviceLocation is one of the PluginConfig fields which has to be a primitive type String. And as I know the PluginConfig fields are populated by reflection rather than a constructor. Therefore, this getter is the first chance to convert the String serviceLocation to URI.. Or we rolled back to the original solution, throw the URISyntaxException in the setter (without throwing a RuntimeException), does it make more sense?

Could you please advice on this? Thanks!

albertshau replied:
I see. In that case it would be cleaner not to have the PluginConfig implement this interface, but to have an extra step where the AnaplanConfig is created from the PluginConfig (with the AnaplanConfig constructor taking URIs for these locations and not Strings).

niceyang replied:
Thanks! That's a better idea! The solution is update!

niceyang committed:
Change the AnaplanConfig from Interface to an entity class.

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -38,15 +38,37 @@
-    URI getServiceLocation() throws URISyntaxException;
+  /** Wrapper for the metadata for the connection with Anaplan server. */
+  public static class AnaplanConfig {
+    private URI serviceLocation;
```
private final for these

niceyang replied:
oh right, updated！

albertshau commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -38,15 +38,37 @@
 
-    URI getAuthServiceLocation() throws URISyntaxException;
+    public AnaplanConfig(
+      String username, String password, String serviceLocation, String authServiceLocation)
```
Generally good to avoid throwing exceptions in a constructor. Here, it's better to pass in the location variables as URI.

We'll get to it in the plugin PR, but we try to capture multiple validation errors at the same time and associate it with the problematic field. Parsing both here loses some context, as you don't know if the exception is because the service location is invalid or if the auth service location is invalid.

niceyang replied:
Thanks for the another instruction! updated!

Yeah, the concern makes sense. We can talk more in the Plugin PR. But started from here ->
For example, in the action plugin, we will call `config.validate(collector)` in configurePipeline(). Does it makes sense if we also run the validate() in the beginning of the `run()`? I dont see this pattern in the existing plugins, but if we do that, the fields will be validated with the context separately. But, we still does not solve the context losing issue in the scope of the AnaplanConfig. Do we much have it as well?
Thanks! :)

albertshau reviewed (approved):
minor comments, otherwise lgtm. please fix and squash commits, then we'll merge.

niceyang committed:
1. Update the fields in the AnaplanConfig to be final.
2. Update parameter type of serviceLocation and authServiceLocation in the AnaplanConfig constructor to be URI.
